src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -0,0 +1,183 @@
+#include <boost/algorithm/string/predicate.hpp>
+#include <chrono>
+#include <fmt/core.h>
+#include <memory>
+#include <thread>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "Common/Exception.h"
+#include "Common/logger_useful.h"
+#include "Checkpoint/CheckpointContext.h"
+#include "Checkpoint/CheckpointContextFwd.h"
+#include "Core/Block.h"
+#include "Formats/FormatFactory.h"
+#include "Interpreters/Context.h"
+#include "KafkaLog/KafkaWALCommon.h"
+#include "KafkaSink.h"
+#include "Processors/Chunk.h"
+#include "Processors/Formats/IOutputFormat.h"
+#include "Processors/ProcessorID.h"
+#include "Storages/ExternalStream/ExternalStreamTypes.h"
+#include "Storages/ExternalStream/Kafka/Kafka.h"
+#include "Storages/ExternalStream/Kafka/WriteBufferFromKafka.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int MISSING_ACKNOWLEDGEMENT;
+extern const int INVALID_CONFIG_PARAMETER;
+}
+
+KafkaSink::KafkaSink(const Kafka * kafka, const Block & header, ContextPtr context, const Poco::Logger * logger)
+    : SinkToStorage(header, ProcessorID::ExternalTableDataSinkID), log(logger), producer(nullptr, nullptr)
+{
+    is_streaming = true;
+
+    /// default values
+    std::vector<std::pair<std::string, std::string>> producer_params{
```

Comment:
I feel like there are quite a few boilplate code which are dup with `KafkaLog`, can we try to modulaize / extend KafkaLog to see if we can have a centralized place to maintain all of this ?

Reply:
Yup, things can be modulized are:
* configuration handling
* error handling

I already tried use existing code in `klog` where it's possible, but there could be some improvement we can do. I think we can make such changes later when I do more work on the source side so that I will get better ideas what can be done.